datafusion/expr/src/expr_schema.rs

```diff
@@ -150,21 +150,22 @@ impl ExprSchemable for Expr {
                     .collect::<Result<Vec<_>>>()?;
 
                 // verify that function is invoked with correct number and type of arguments as defined in `TypeSignature`
-                data_types_with_scalar_udf(&arg_data_types, func).map_err(|err| {
-                    plan_datafusion_err!(
-                        "{} {}",
-                        err,
-                        utils::generate_signature_error_msg(
-                            func.name(),
-                            func.signature().clone(),
-                            &arg_data_types,
+                let new_data_types = data_types_with_scalar_udf(&arg_data_types, func)
```

Comment:
I think this is the root cause of the issue and to solve this other changes are necessary. Therefore, I think we should go with this change and maybe further optimize the coercion in another PR.

Reply:
Ok, so should I leave it as it is? Or change it back to how it was:

```rust
data_types_with_scalar_udf(&arg_data_types, func)
```

Reply:
I'm ok with the current change, maybe wait for @findepi @alamb for how to move on with this PR

Reply:
I defer to @jayzhan211  -- if he is good to merge this PR, let's get the conflicts resolved and merge it in.

If there is additional work we know is needed / could be cleaned up, let's try and file them as tickets

Reply:
Conflicts solved! :smile: